### DISCUSSION TITLE
Best place to start Avalonia for beginner

### DISCUSSION THREAD

**Discussion by kodemetrix** [AvaloniaUI/Avalonia#6881] [Q&A]
Hi All,

newbie here!

Could anyone advice me where would be the best place to start learning Avalonia from? obviously, I have no WPF experience but WinForms and C#...

Thanks!

**Comment by jmacato** [comment]
I think this tutorial is a good place to start :) 
https://docs.avaloniaui.net/tutorials/todo-list-app

**Comment by rstm-sf** [comment]
If you have no experience with MVVM, or rather with the use of [INPC](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.inotifypropertychanged), then it is recommended that you familiarize yourself with it first. This can be done with winforms as well. For example, [the project](https://github.com/oxyplot/oxyplot/tree/develop/Source/Examples/WindowsForms/ExampleBrowser).

Then you can complete the tutorial here, which was thrown off above. It is recommended not to use ReactiveUI unless you have experience implementing INPC. This will make it easier to understand

After that, it will be easier to understand what to do, based on the tutorials on WPF

**Comment by npolyak** [comment]
I have a bunch of Avalonia articles for those who do not know WPF on the codeproject - 

- [Multiplatform UI Coding with AvaloniaUI in Easy Samples. Part 1 - AvaloniaUI Building Blocks](https://www.codeproject.com/Articles/5308645/Multiplatform-UI-Coding-with-AvaloniaUI-in-Easy-Sa)
- [Basics of XAML in Easy Samples for Multiplatform Avalonia .NET Framework](https://www.codeproject.com/Articles/5314369/Basics-of-XAML-in-Easy-Samples-for-Multiplatform-A)
- [Multiplatform Avalonia .NET Framework Programming Basic Concepts in Easy Samples](https://www.codeproject.com/Articles/5311995/Multiplatform-Avalonia-NET-Framework-Programming-B)
- [Multiplatform Avalonia .NET Framework Programming Advanced Concepts in Easy Samples](https://www.codeproject.com/Articles/5317055/Multiplatform-Avalonia-NET-Framework-Programming-A)

Each article has code samples. You should also go through the code samples while reading the articles. 

If you read the articles, please, tell me what you think can be improved. 

Thanks

**Reply by donnyv** [reply]
First I want to say thank you for doing these articles. This is exactly what I've been looking for. The Avalonia has so many ways to do the same thing that I was getting very frustrated on how things worked and trade offs with the different options. Do you have an article on using ReactiveUI with Avalonia?

**Reply by npolyak** [reply]
Thanks Donny. Unfortunately I do not have such article yet. ReactiveUI is not something I've been using a lot myself preferring to build the logical and visual trees within XAML (I know a lot of people prefer using C# for that, but I think that XAML is good for representing the trees). Perhaps in the future I'll research ReactiveUI and start using it. My first goal, however is to master DynamicData (also an Rx based package) which allows to build some very interesting and powerful View Models absorbing some of the complexity usually placed in the UI, e.g. Paging, Caching, Virtualization and Hierarchy.

**Reply by donnyv** [reply]
So by using ReactiveUI, is that the only way to make the rendering asynchronous?

**Reply by npolyak** [reply]
Rendering is a different thread that we do not have control of. Composition of the Visual Tree should always be done on the dispatcher thread whether using ReactiveUI or not. The only things that can be run asynchronously are things that do not directly affect the UI, e.g. long calls to the backend or some complex calculations on the view model.

**Comment by anst-foto** [comment]
Есть небольшой видеокурс по основам Avalonia на русском языке - [StarIT. Avalonia](https://github.com/StarIT-AnSt/courses/blob/main/avalonia.md)

**Comment by mysteryx93** [comment]
Here's a tutorial I wrote
[Modern.Net-Tutorial](https://github.com/mysteryx93/Modern.Net-Tutorial)
